### PR TITLE
Frontend: Migrate DataSource HTTP Settings to React

### LINKED ISSUE #16857: Migrate datasource HTTP settings to react
In #16307, we made it possible to have a react ConfigEditor -- it works great, but you can't change the HTTP settings.

We need to expose the settings from:
https://github.com/grafana/grafana/blob/master/public/app/features/datasources/partials/http_settings.html

### PR DESCRIPTION
**Pull request by dprokop**
Closes https://github.com/grafana/grafana/issues/16857

TODO: 
- [x] migrate `bootstrap-tagsinput`
- [x] add validation to URL field
- [x] add `secureJsonFields` to DataSourceSettings

#### Decide on whether or not make this new component backward compatible with <datasource-http-settings> directive
It can be used by external plugins so we need to be thought here. Not sure what's the best approach here but it feels like having `https://github.com/grafana/grafana/pull/19452/files#diff-3fb435339dc77549d6f7e3fab49adc68R13` props on a new component is not correct. Maybe we could have <datasource-http-settings-next .../> that would be react2angular wrapper, with a new API, and we could refactor the existing datasource-http-settings directive to use the new one in template?

### CODE REVIEW

**Comment by dprokop** [comment]
@peterholmberg I think we gonna be fine about backward compatibility with this commit: ae9ba497ba8890ec475c310c771a403793e61cdf

File: public/app/core/angular_wrappers.ts [hunk 1]
```diff
@@ -109,4 +115,10 @@ export function registerAngularDirectives() {
     'onChange',
     ['datasource', { watchDepth: 'reference' }],
   ]);
+  react2AngularDirective('datasourceHttpSettingsNext', DatasourceHttpSettings, [
```
Comment (torkelo) [inline public/app/core/angular_wrappers.ts]: we need something that is backward compatible with the old directive as well.

**Comment by torkelo** [comment]
It looks close now! 

image

think the button should be the secondary color not primary. Would be great if the tags showed up inside the input (to the left of the real input field).

Another thing I noticed is that when I click Add the whole page is saved & datasource tested. This should only happen when you click "Save & Test" at the bottom not click add.

**Comment by torkelo** [comment]
Is this ready for more review/testing?

File: packages/grafana-ui/src/components/DataSourceSettings/TLSAuthSettings.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
+import React from 'react';
+import { HttpSettingsBaseProps } from './types';
+import { KeyValue } from '@grafana/data';
+
+// TODO: Refactor forms below to a reusable Cert component
+export const TLSAuthSettings: React.FC<HttpSettingsBaseProps> = ({ dataSourceConfig, onChange }) => {
+  console.log();
```
Comment (dprokop) [inline packages/grafana-ui/src/components/DataSourceSettings/TLSAuthSettings.tsx]:
```suggestion
```

File: packages/grafana-ui/src/components/DataSourceSettings/TLSAuthSettings.tsx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  const hasTLSCACert = dataSourceConfig.secureJsonFields && dataSourceConfig.secureJsonFields.tlsCACert;
+  const hasTLSClientCert = dataSourceConfig.secureJsonFields && dataSourceConfig.secureJsonFields.tlsClientCert;
+  // @ts-ignre
```
Comment (dprokop) [inline packages/grafana-ui/src/components/DataSourceSettings/TLSAuthSettings.tsx]:
```suggestion
```

File: packages/grafana-ui/src/components/DataSourceSettings/TLSAuthSettings.tsx (hunk 2 above)
Comment (peterholmberg) [reply packages/grafana-ui/src/components/DataSourceSettings/TLSAuthSettings.tsx]: Completely missed this feature

File: packages/grafana-ui/src/components/DataSourceSettings/TLSAuthSettings.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      ...dataSourceConfig,
+      secureJsonData: newSecureJsonData,
+    });
+  };
+
+  return (
+    <div className="gf-form-group">
+      <div className="gf-form">
+        <h6>TLS Auth Details</h6>
+        <h6>TODO: add missing info popover</h6>
```
Comment (dprokop) [inline packages/grafana-ui/src/components/DataSourceSettings/TLSAuthSettings.tsx]: Minor: missing info icon
image

File: public/app/features/datasources/settings/HttpSettingsCtrl.ts [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import { coreModule } from 'app/core/core';
-import { createChangeHandler, createResetHandler, PasswordFieldEnum } from '../utils/passwordHandlers';
+// import { createChangeHandler, createResetHandler, PasswordFieldEnum } from '../utils/passwordHandlers';
```
Comment (dprokop) [inline public/app/features/datasources/settings/HttpSettingsCtrl.ts]:
```suggestion
```

File: public/app/core/angular_wrappers.ts (hunk 1 above)
Comment (dprokop) [reply public/app/core/angular_wrappers.ts]: @torkelo for the backward compatibility I have not removed the old directive. Instead, it's using the new one under the hood https://github.com/grafana/grafana/pull/19452/files#diff-11d01ebf4e4e29f39faeca6e1f9273f6R4

Comment (torkelo) [reply public/app/core/angular_wrappers.ts]: yes, that's great. I missed that

**Comment by dprokop** [comment]
@peterholmberg ref my comment above, here's update styling.md readme: https://github.com/grafana/grafana/blob/master/contribute/style-guides/styling.md#styling-complex-components

File: public/app/features/datasources/settings/HttpSettingsCtrl.ts (hunk 5 above)
Comment (peterholmberg) [reply public/app/features/datasources/settings/HttpSettingsCtrl.ts]: should be able to remove other commented code in this file too?

Comment (dprokop) [reply public/app/features/datasources/settings/HttpSettingsCtrl.ts]: yes

**Comment by peterholmberg** [comment]
@torkelo @dprokop could you give this a third round of review?

**Comment by dprokop** [comment]
I got this warning when switching the `Forward OAuth Identity` option

image

**Comment by dprokop** [comment]
Once you configure basic auth credentials, they are never cleaned up if you disable basic auth. To repro:

1. Enable basic auth and add username/password
2. Disable basic auth
2. Enable basic auth back - previous credentials are still there

**Comment by torkelo** [comment]
I think if we save with basic auth disabled we should clear credentials, same with CA & TLS. 
But only on save (so that just accidentally clicking on the checkbox should not clear it)

**Review by torkelo — APPROVED** [review]
Looks good! Great work! 

One thing we are missing is for a data source to use this directly from the react ConfigEditor. Will create a follow-up issue to migrate our internal data source config's to 100% react.

File: packages/grafana-ui/src/components/DataSourceSettings/DataSourceHttpSettings.tsx [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  const accessSelect = (
+    <Select
+      width={20}
+      options={ACCESS_OPTIONS}
+      value={ACCESS_OPTIONS.filter(o => o.value === dataSourceConfig.access)[0] || DEFAULT_ACCESS_OPTION}
+      onChange={selectedValue => onSettingsChange({ access: selectedValue.value })}
+    />
+  );
+
+  const isValidUrl = /^(ftp|http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?$/.test(
```
Comment (aocenas) [inline packages/grafana-ui/src/components/DataSourceSettings/DataSourceHttpSettings.tsx]: @dprokop is there a reason not to use `validationEvents` prop on Inputs? Should that be in general used or is there some issues with that?

**Comment by aocenas** [comment]
Also `FormField` vs using `FormLabel` + some input, what is the preferred one? Seems here both are used and I don't see any functional difference.

File: packages/grafana-ui/src/components/DataSourceSettings/DataSourceHttpSettings.tsx (hunk 6 above)
Comment (dprokop) [reply packages/grafana-ui/src/components/DataSourceSettings/DataSourceHttpSettings.tsx]: @aocenas good question. I think we could use validationEvents in this case, can't see a reason why not. @peterholmberg do you wanna try it out?

**Comment by dprokop** [comment]
@aocenas I would suggest using FormLabel + input as it will be easier to migrate to new form styles in the future (as those are priitives that we will be able to easily replace one place after another).

**Comment by aocenas** [comment]
@peterholmberg @dprokop the Tags input needs some better styling for light theme:
Screenshot 2019-10-18 at 15 47 17

**Pull request merged**